js/modules/k6/experimental/fs/module.go

```diff
@@ -0,0 +1,185 @@
+// Package fs provides a k6 module that allows users to interact with files from the
+// local filesystem.
+package fs
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/dop251/goja"
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/js/promises"
+	"go.k6.io/k6/lib/fsext"
+)
+
+type (
+	// RootModule is the global module instance that will create instances of our
+	// module for each VU.
+	RootModule struct {
+		fileRegistry fileRegistry
```

Comment:
I do wonder if this even makes sense from design/implementation perspective.

Technically we want to have as few copies of the file data as possible. This in practice adds 1, so uh not great.

All the methods that in practice will not copy the whole body (which will need to have a complete copy either way) are in practice also implemented on the current file system abstraction (afero.FS) - seek, read, stat, close.

io/fs is not guraunteed to implement `seek` it seems, but it likely will for all of *our* cases.

So I do wonder if we even need this to begin with instead of directly using the underlying fs abstraction and wrap them in js.

Am I missing a case?

Reply:
I think I understand where you're coming from: why do we cache the data when we could access it through the fs (or afero) library directly? Correct?

The initial main reason was: because there are limited file descriptors available to the OS, and in the event, we have a number of VUs > number of available file descriptors, the OS is going to start refusing our requests to open files.

But now that you say it, as I write this, I also come to realize that we actually open a file descriptor once, store it in the registry, and then have various `File` instances with their own internal offsets as we do now, but instead, later call `ReadAt` on them for instance. I'm gonna give that a shot 👍🏻

Reply:
> The initial main reason was: because there are limited file descriptors available to the OS

That is *true*. In the last 10+ years there is usually no problems to have hundreds of thousands if not even millions of file descriptors. The actual problem AFAIK for a long time has been that `select` has a limit of 1024 so if you have a big limit globally and a program with `select` it can panic in a ... bad way.

But in practice k6 has to be run with more than 1024 fds as it also does network requests and it does quite a lot of those. We do not share those between VUs for some reason, and I don't think we should do it for files.

Most use cases and users are unlikely to be running so many VUs on a single machine and opening so many files that they will hit the limit. And if they *do* - I would argue raising the limit will work way better than us having a lock and basically synchronizing access to the fs. 

I would expect that even at 1k VUs trying to read 100b each sequentially it will practically spend most of it in locking and unlocking.

This is most based on

> but instead, later call ReadAt on them for instance

Where it seems you want to `ReadAt` which seems to be race safe from reading the code - although I would argue the documentation doesn't make it as explicit as I would like. 

So maybe we can use it ... maybe we should just directly "proxy" to `Read` and let the OS (in the future when afero.MemMapsfs isn't the actual backing) handle it.

Reply:
I don't understand what you're suggesting we should do then? Can you clarify that?

_some context I have_: 
* my understanding is that Go reuses underlying TCP connections (and thus file descriptors?) under the hood, under normal conditions (not forgetting to close the response body), thus limiting the number of file descriptors opened at the same time, even when firing tons of requests concurrently. But I guess you'll have more details on this.
* my (just now) empirical tests show that linux has a limit of 1024 open fd (on my ubuntu server) per process, macos seems to be somewhere around 10k, and windows is much much farther. 
* changing the limit on linux is somewhat easy, on mac it seems possible, on windows I couldn't find much conclusive. All in all, telling users "you can open files, but you're gonna have to fine tune your machine" is not really a good user experience.
* the time spent locking/unlocking is good point indeed
* I've made some tests with readAt up to 10k concurrent readers and no race detected

Reply:
> my (just now) empirical tests show that linux has a limit of 1024 open fd (on my ubuntu server) per process, macos seems to be somewhere around 10k, and windows is much much farther.

> changing the limit on linux is somewhat easy, on mac it seems possible, on windows I couldn't find much conclusive. All in all, telling users "you can open files, but you're gonna have to fine tune your machine" is not really a good user experience.

Both of those are valid and will have the same problem if you do 2000 network requests and your limit is 1000. 

So making it so that opening 1000 files is any different than making 1000 network requests seems like not needed optimization.

> I've made some tests with readAt up to 10k concurrent readers and no race detected

As I mention above - it seems like it will be (at least on Posix) from my reading of the code as it uses a syscall that does ReadAt in one go instead of doing as Seek +  Read. 

In our current underlying fs(afero.fs) - this is even better. The whole file is in memory already so `ReadAt` is reading parts of a slice essentially. 

My concerns:
-  ~I am not certain ReadAt is actually supposed to be safe for concurrent use~ Reading the docs on [`io.ReaderAt`](https://pkg.go.dev/io#ReaderAt) it is defined it is supposed to be safe to call it in parallel.
-  `io/fs` (which we would like to move to) is not guaranteed to have it implemented and implementing for some case might be harder than expected.

So keeping with `Read` and multiple files seems like the better option. 

Also, from implementation perspective having a registry means that we have to keep count of it, lock it at least on `open`. I guess we might also might want to remove stuff from it on `close` if all VUs have closed it :shrug:.

Not having one just removes all of those questions and needed considerations.
 
tl;dr: IMO having a registry is not worth it in terms of what it adds as features, but requires us to make more complicated implementation. But as with most code (especially one in experimental modules) I doubt it will stay the same long term, so I don't think it is that big of a problem having it as well. I am just mostly not sold on the benefits from it.

Reply:
Thanks for your writeup 🙇🏻 

I've been giving some thoughts to it, and experimented with how I understood the alternative you propose. 

For this first round of implementation of the module, I'd rather stick to the current implementation, and would be very keen to revise it at a later point in time indeed (#3020 and #2978 for instance).